src/Gaufrette/Adapter/Local.php

```diff
@@ -130,20 +138,34 @@ public function mtime($key)
     /**
      * {@inheritdoc}
      *
+     * Can also delete a directory recursively when the given $key matches a
+     * directory.
      */
     public function delete($key)
     {
         if ($this->isDirectory($key)) {
-            if (!rmdir($this->computePath($key))) {
+            try {
+                if (!$this->deleteDirectory($this->computePath($key))) {
+                    throw StorageFailure::unexpectedFailure('delete', ['key' => $key]);
+                }
+            } catch (\InvalidArgumentException $e) {
+                throw StorageFailure::unexpectedFailure('delete', ['key' => $key], $e);
+            }
+
+            return;
+        } elseif ($this->exists($key)) {
+            if (!unlink($this->computePath($key))) {
                 throw StorageFailure::unexpectedFailure('delete', ['key' => $key]);
             }
 
             return;
         }
 
-        if (!unlink($this->computePath($key))) {
-            throw StorageFailure::unexpectedFailure('delete', ['key' => $key]);
-        }
+        throw StorageFailure::unexpectedFailure(
+            'delete',
+            ['key' => $key],
+            new FileNotFound($key)
```

Comment:
You don't have to wrap `FileNotFound` exception in a `StorageFailure` actually. Both are valid exceptions accordingly to the Adapter interface (see [here](https://sourcegraph.com/github.com/KnpLabs/Gaufrette@1.x/-/blob/src/Gaufrette/Adapter.php#L84)). `StorageFailure` should be thrown only if others, more specific exceptions can't be thrown :wink: